Kitodo-LongTimePreservationValidationModule/src/main/java/org/kitodo/longtimepreservationvalidationmodule/ValidationResultState.java

```diff
@@ -13,52 +13,113 @@
 
 import edu.harvard.hul.ois.jhove.RepInfo;
 
+import java.util.Optional;
+
 import org.kitodo.api.validation.State;
+import org.kitodo.config.KitodoConfig;
+
+/**
+ * The validation result state is used to map the two-valued result state of
+ * JHove to the Kitodo validation interface.
+ */
+public class ValidationResultState {
+
+    /**
+     * The constant for the indefinite state.
+     */
+    private static final String UNDETERMINED = "UNDETERMINED";
 
-public enum ValidationResultState {
-    ERROR(State.ERROR),
-    NOT_WELL_FORMED(State.ERROR),
-    UNKNOWN(State.WARNING),
-    VALID(State.SUCCESS),
-    WELL_FORMED(State.SUCCESS),
-    WELL_FORMED_BUT_NOT_VALID(State.WARNING);
+    /**
+     * Constant for the error case.
+     */
+    public static final ValidationResultState ERROR = new ValidationResultState() {
+        @Override
+        public State toState() {
+            return State.ERROR;
+        }
+    };
 
     /**
-     * Returns the enum constant for the specified wellformedness and valididity
-     * values.
+     * Saves the value of well-formedness.
+     */
+    private Optional<Boolean> wellFormed;
```

Comment:
JHove returns three different values: true, false or "I do not know". I wanted to show that to something meaningful and that came out of it.

Reply:
Boolean also has 3 values: true, false and null.

Reply:
You're right. I have often thought about it. But if you then try on the object with the three states, e.g. to call `.toString()` you will not get "true", "false" and "null", but "true", "false" and a NullPointerException. The same with .equals(). That's why I do not do that. `null` is always something undefined, while we know here that we do not know anything. It does not look much better at first glance, but it makes a huge difference. By the way [there is a nice Wikipedia page to it](https://en.wikipedia.org/wiki/There_are_known_knowns). It is not a waste of time to read.

Reply:
I understand that you prefer new fancy ways of programming but I would start from learn how to actually use properly Optional, before that I advice to use old reliable methods with good boring check if value is not null.

![kitodo](https://user-images.githubusercontent.com/9614459/46790088-6458ae80-cd3e-11e8-8086-d8b01995bfa8.png)

Reply:
I've been bumped into this several times that I'm not allowed to use null objects in the Kitodo project and that I should use Optional. In the beginning, I also thought that was superfluous, but meanwhile I got used to it, and maybe it has its good side. If this has changed again and I should use `null` references from now on, from me. But then please write it down somewhere and document it so that our approach does not change every few months.

The bug list you are attaching is really correct. For Optional, the pattern is always to check first if the object is there before using it. Otherwise you get nothing else than a `NullPointerException`, and that's exactly what the object should avoid. But probably those are where the value is simply needed, and if it is not there you could do nothing but throw an exception. That would have to look at the individual places, also see how talking the exception is and whether it makes sense to check each time and throw an exception by hand, if the code otherwise throws a meaningful exception automatically. Can you open a ticket for that?

Reply:
Best with line numbers for the bugs?

Reply:
I would use an enum class for this, but this is only my personal way if I have more then two values to return from a method and a boolean value is not enough.

Reply:
@matthias-ronge Using null return is wrong but using Optional without checks (it is clearly what you do as SonarCloud warns about it in almost every place where you use it) is also wrong. In general there is also other problem with Optional. It has very bad performance (https://stackoverflow.com/questions/34696884/performance-of-java-optional).

> Comparison of two several runs with different size of the list shows that approach with code that uses Optional at the best is twice slower than one with nullables. With small lists it is 3 times slower.

@henning-gerhardt It was my first remark that this class was enum and now is changed to class wich uses Optional.

Reply:
Although I find it completely superfluous to introduce my own class, but I did it. I hope it is good now.

Reply:
PS: I just looked it up: An `Optional` throws, if there is no value, a `NoSuchElementException` with the text "No value present". Can one now discuss whether it is completely wrong that a function that needs a value throws a `NoSuchElementException` if it does not get one. If that's not enough, we actually have to do the review at each point and throw an exception by hand.